Mathlib/Order/Rel/GaloisConnection.lean

```diff
@@ -0,0 +1,108 @@
+  Laurent Lafforgue (in preparation)
+
+## Tags
+
+relation, Galois connection, induced bijection, fixed points
+-/
+
+section
+
+universe u v
+
+variable {α : Type u} {β : Type v} (R : Rel α β)
+
+namespace Rel
+
+/-! ### Pairs of adjoint maps defined by relations -/
+
+open OrderDual
+
+/-- `leftDual` maps any set `J` of elements of type `α` to the set `{b : β | ∀ a ∈ J, R a b}` of
+elements `b` of type `β` such that `R a b` for every element `a` of `J`. -/
+def leftDual (J : Set α) : Set β := {b : β | ∀ ⦃a⦄, a ∈ J → R a b}
+
+/-- `rightDual` maps any set `I` of elements of type `β` to the set `{a : α | ∀ b ∈ I, R a b}`
+of elements `a` of type `α` such that `R a b` for every element `b` of `I`. -/
+def rightDual (I : Set β) : Set α := {a : α | ∀ ⦃b⦄, b ∈ I → R a b}
+
+/-- The pair of functions `toDual ∘ leftDual` and `rightDual ∘ ofDual` forms a Galois connection. -/
+theorem gc_leftDual_rightDual : GaloisConnection (toDual ∘ R.leftDual) (R.rightDual ∘ ofDual) :=
+  fun J I ↦ ⟨fun h a ha b hb ↦ h (by simpa) ha, fun h b hb a ha ↦ h (by simpa) hb⟩
+
+/-! ### Induced equivalences between fixed points -/
+
+/-- `leftFixedPoints` is the set of elements `J : Set α` satisfying `rightDual (leftDual J) = J`. -/
+def leftFixedPoints := {J : Set α | R.rightDual (R.leftDual J) = J}
+
+/-- `rightFixedPoints` is the set of elements `I : Set β` satisfying `leftDual (rightDual I) = I`. -/
+def rightFixedPoints := {I : Set β | R.leftDual (R.rightDual I) = I}
+
+open GaloisConnection
+
+/-- `leftDual` maps every element `J` to `rightFixedPoints`. -/
+theorem leftDual_mem_rightFixedPoint (J : Set α) : R.leftDual J ∈ R.rightFixedPoints := by
+  apply le_antisymm
+  · apply R.gc_leftDual_rightDual.monotone_l; exact R.gc_leftDual_rightDual.le_u_l J
+  · exact R.gc_leftDual_rightDual.l_u_le (R.leftDual J)
+
+/-- `rightDual` maps every element `I` to `leftFixedPoints`. -/
+theorem rightDual_mem_leftFixedPoint (I : Set β) : R.rightDual I ∈ R.leftFixedPoints := by
+  apply le_antisymm
+  · apply R.gc_leftDual_rightDual.monotone_u; exact R.gc_leftDual_rightDual.l_u_le I
+  · exact R.gc_leftDual_rightDual.le_u_l (R.rightDual I)
+
+/-- The maps `leftDual` and `rightDual` induce inverse bijections between the sets of fixed points.
+-/
+def equivFixedPoints : R.leftFixedPoints ≃ R.rightFixedPoints where
+  toFun := fun ⟨J, _⟩ => ⟨R.leftDual J, R.leftDual_mem_rightFixedPoint J⟩
+  invFun := fun ⟨I, _⟩ => ⟨R.rightDual I, R.rightDual_mem_leftFixedPoint I⟩
+  left_inv J := by cases' J with J hJ; simp; rw [hJ]
+  right_inv I := by cases' I with I hI; simp; rw [hI]
+
+theorem rightDual_leftDual_le_of_le {J J' : Set α} (h : J' ∈ R.leftFixedPoints) :
+    J ≤ J' → R.rightDual (R.leftDual J) ≤ J' := by
+  intro h₁
+  rw [← h]
+  apply R.gc_leftDual_rightDual.monotone_u
+  apply R.gc_leftDual_rightDual.monotone_l
+  exact h₁
+
+theorem ge_leftDual_rightDual_of_ge {I I' : Set β} (h : I' ∈ R.rightFixedPoints) :
+    I' ≥ I → I' ≥ R.leftDual (R.rightDual I) := by
```

Comment:
Please never `≥` but always `≤`. We don't want people having to guess two names and two statements
```suggestion
theorem leftDual_rightDual_le_of_le {I I' : Set β} (h : I' ∈ R.rightFixedPoints) :
    I ≤ I' → R.leftDual (R.rightDual I) ≤ I' := by
```
also you might as well pull out the assumption to before the colon
```suggestion
theorem leftDual_rightDual_le_of_le {I I' : Set β} (h : I' ∈ R.rightFixedPoints) (h₁ : I ≤ I') :
    R.leftDual (R.rightDual I) ≤ I' := by
```

Reply:
> Please never `≥` but always `≤`. We don't want people having to guess two names and two statements

Make sense!